### PR TITLE
[flutter_tools] add custom tool analysis to analyze.dart, lint Future.catchError

### PR DESCRIPTION
**Pull request by christopherfujino**
Ensure tool code does not use Future.catchError or Future.onError, because it is not statically safe: https://github.com/dart-lang/sdk/issues/51248.

This was proposed upstream in dart-lang/linter in https://github.com/dart-lang/sdk/issues/59040 and https://github.com/dart-lang/linter/pull/4068, but not accepted.

### CODE REVIEW

**Comment by christopherfujino** [comment]
This is still on my radar

**Review by andrewkolos — APPROVED** [review]
LGTM.

Nit for future reference: The title struck me as non-sequitur to the description until I read the diff.

Title could have been something like "Add lint for X." The description would explain 1) why we want this lint and then 2) that we first need to give the custom CI analyzer code the ability to run rules against only the tool codebase, hence the big change in analyze.dart.

**Comment by christopherfujino** [comment]
> LGTM.
> 
> Nit for future reference: The title struck me as non-sequitur to the description until I read the diff.
> 
> Title could have been something like "Add lint for X." The description would explain 1) why we want this lint and then 2) that we first need to give the custom CI analyzer code the ability to run rules against only the tool codebase, hence the big change in analyze.dart.

Good catch, updated the title.

File: dev/bots/custom_rules/no_double_clamp.dart [hunk 1]
```diff
@@ -39,14 +38,10 @@ class _NoDoubleClamp implements AnalyzeRule {
       return;
     }
 
-    String locationInFile(ResolvedUnitResult unit, AstNode node) {
```
Comment (christopherfujino) [inline dev/bots/custom_rules/no_double_clamp.dart:42]: oops, this looks like a bad merge

Comment (christopherfujino) [reply dev/bots/custom_rules/no_double_clamp.dart:42]: nvm, i moved this to utils so multiple rules could share

**Pull request merged**